File: python/algorithms/greedy/greedy.py
"""贪心：活动选择、分数背包。"""

from __future__ import annotations


def activity_selection(intervals: list[tuple[int, int]]) -> int:
    """intervals: (start, end)，最多不相交区间数。"""
    intervals = sorted(intervals, key=lambda x: x[1])
    cnt = 0
    last_end = -10**9
    for s, e in intervals:
        if s >= last_end:
            cnt += 1
            last_end = e
    return cnt


def fractional_knapsack(items: list[tuple[int, int]], capacity: int) -> float:
    """items: (weight, value)，可分割，求最大价值。"""
    ratio = [(v / w, w, v) for w, v in items if w > 0]
    ratio.sort(reverse=True)
    rem = float(capacity)
    total = 0.0
    for _r, w, v in ratio:
        if rem <= 0:
            break
        take = min(rem, w)
        total += take * (v / w)
        rem -= take
    return total


if __name__ == "__main__":
    assert activity_selection([(1, 2), (3, 4), (5, 6), (7, 8)]) == 4
    assert abs(fractional_knapsack([(10, 60), (20, 100), (30, 120)], 50) - 240.0) < 1e-9
    print("greedy OK")

File: cpp/algorithms/greedy/greedy.cpp
#include <bits/stdc++.h>
using namespace std;

int activity_selection(vector<pair<int, int>> iv) {
    sort(iv.begin(), iv.end(), [](auto& a, auto& b) { return a.second < b.second; });
    int cnt = 0, last = INT_MIN / 2;
    for (auto [s, e] : iv)
        if (s >= last) {
            ++cnt;
            last = e;
        }
    return cnt;
}

double fractional_knapsack(vector<pair<int, int>> items, int cap) {
    vector<tuple<double, int, int>> r;
    for (auto [w, v] : items)
        if (w > 0) r.push_back({(double)v / w, w, v});
    sort(r.begin(), r.end(), greater<tuple<double, int, int>>());
    double rem = cap, tot = 0;
    for (auto [rv, w, v] : r) {
        if (rem <= 0) break;
        double take = min(rem, (double)w);
        tot += take * ((double)v / w);
        rem -= take;
    }
    return tot;
}

int main() {
    assert(activity_selection({{1, 2}, {3, 4}, {5, 6}, {7, 8}}) == 4);
    assert(abs(fractional_knapsack({{10, 60}, {20, 100}, {30, 120}}, 50) - 240.0) < 1e-9);
    cout << "greedy OK" << endl;
    return 0;
}
